File: Samples/JumpList/js/js/scenario4-changesystemgroup.js
//// Copyright (c) Microsoft Corporation. All rights reserved

(function () {
    "use strict";

    var ApplicationData = Windows.Storage.ApplicationData;
    var StorageApplicationPermissions = Windows.Storage.AccessCache.StorageApplicationPermissions;
    var RecentStorageItemVisibility = Windows.Storage.AccessCache.RecentStorageItemVisibility;
    var StartScreen = Windows.UI.StartScreen;
    var JumpList = StartScreen.JumpList;
    var JumpListSystemGroupKind = StartScreen.JumpListSystemGroupKind;

    var page = WinJS.UI.Pages.define("/html/scenario4-changesystemgroup.html", {
        ready: function (element, options) {
            SystemGroup_Recent.addEventListener("change", onSystemGroupRecentChange);
            SystemGroup_Frequent.addEventListener("change", onSystemGroupFrequentChange);
            SystemGroup_None.addEventListener("change", onSystemGroupNoneChange);
            PrepareSampleFiles.addEventListener("click", onPrepareSampleFilesClick);
            loadSavedSystemGroup();
        }
    });

    function onSystemGroupRecentChange() {
        if (SystemGroup_Recent.checked) {
            JumpList.loadCurrentAsync().done(function (jumpList) {
                jumpList.systemGroupKind = JumpListSystemGroupKind.recent;
                jumpList.saveAsync();
            });
        }
    }

    function onSystemGroupFrequentChange() {
        if (SystemGroup_Frequent.checked) {
            JumpList.loadCurrentAsync().done(function (jumpList) {
                jumpList.systemGroupKind = JumpListSystemGroupKind.frequent;
                jumpList.saveAsync();
            });
        }
    }

    function onSystemGroupNoneChange() {
        if (SystemGroup_None.checked) {
            JumpList.loadCurrentAsync().done(function (jumpList) {
                jumpList.systemGroupKind = JumpListSystemGroupKind.none;
                jumpList.saveAsync();
            });
        }
    }

    function addFile(fileName)
    {
        var localFolder = ApplicationData.current.localFolder;
        localFolder.tryGetItemAsync(fileName).done(function (file) {
            if (file == null) {
                localFolder.createFileAsync(fileName).done(function (newFile) {
                    StorageApplicationPermissions.mostRecentlyUsedList.add(newFile, "", RecentStorageItemVisibility.appAndSystem);
                });
            }
        });
    }

    function onPrepareSampleFilesClick() {
        // NOTE: This is for the sake of the sample only. Real apps should allow
        // the system list to be populated by user activity.

        for (var i = 0; i < 5; ++i) {
            addFile("Temp" + i + ".ms-jumplist-sample");
        }
    }

    function loadSavedSystemGroup() {
        JumpList.loadCurrentAsync().done(function (jumpList) {
            switch (jumpList.systemGroupKind)
            {
                case JumpListSystemGroupKind.recent:
                    SystemGroup_Recent.checked = true;
                    break;
                case JumpListSystemGroupKind.frequent:
                    SystemGroup_Frequent.checked = true;
                    break;
                case JumpListSystemGroupKind.none:
                    SystemGroup_None.checked = true;
                    break;
            }
        });
    }
})();

File: Samples/JumpList/cppwinrt/Scenario4_ChangeSystemGroup.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "Scenario4_ChangeSystemGroup.h"
#include "Scenario4_ChangeSystemGroup.g.cpp"

using namespace winrt;
using namespace Windows::Foundation;
using namespace Windows::Storage;
using namespace Windows::Storage::AccessCache;
using namespace Windows::UI::StartScreen;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Navigation;

namespace winrt::SDKTemplate::implementation
{
    Scenario4_ChangeSystemGroup::Scenario4_ChangeSystemGroup()
    {
        InitializeComponent();
    }

    fire_and_forget Scenario4_ChangeSystemGroup::OnNavigatedTo(NavigationEventArgs const&)
    {
        auto lifetime = get_strong();

        JumpList jumpList = co_await JumpList::LoadCurrentAsync();
        switch (jumpList.SystemGroupKind())
        {
        case JumpListSystemGroupKind::Recent:
            SystemGroup_Recent().IsChecked(true);
            break;
        case JumpListSystemGroupKind::Frequent:
            SystemGroup_Frequent().IsChecked(true);
            break;
        case JumpListSystemGroupKind::None:
            SystemGroup_None().IsChecked(true);
            break;
        }
    }

    fire_and_forget Scenario4_ChangeSystemGroup::SystemGroup_Recent_Click(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();
        JumpList jumpList = co_await JumpList::LoadCurrentAsync();
        jumpList.SystemGroupKind(JumpListSystemGroupKind::Recent);
        co_await jumpList.SaveAsync();
    }

    fire_and_forget Scenario4_ChangeSystemGroup::SystemGroup_Frequent_Click(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();
        JumpList jumpList = co_await JumpList::LoadCurrentAsync();
        jumpList.SystemGroupKind(JumpListSystemGroupKind::Frequent);
        co_await jumpList.SaveAsync();
    }

    fire_and_forget Scenario4_ChangeSystemGroup::SystemGroup_None_Click(IInspectable const&, RoutedEventArgs const&)
    {
        JumpList jumpList = co_await JumpList::LoadCurrentAsync();
        jumpList.SystemGroupKind(JumpListSystemGroupKind::None);
        co_await jumpList.SaveAsync();
    }

    fire_and_forget Scenario4_ChangeSystemGroup::PrepareSampleFiles_Click(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();

        // NOTE: This is for the sake of the sample only. Real apps should allow
        // the system list to be populated by user activity.

        StorageFolder localFolder = ApplicationData::Current().LocalFolder();

        for (int i = 0; i < 5; ++i)
        {
            hstring fileName{ L"Temp" + std::to_wstring(i) + L".ms-jumplist-sample" };
            StorageFile file = co_await localFolder.CreateFileAsync(fileName, CreationCollisionOption::OpenIfExists);
            StorageApplicationPermissions::MostRecentlyUsedList().Add(file, L"", RecentStorageItemVisibility::AppAndSystem);
        }
    }
}
